### ISSUE TITLE
Join loses sharing, causes OOM

### ISSUE THREAD

**Issue by lukemaurer** [ocaml-flambda/flambda-backend#472]
In the Jane Street tree, we have a large functor (Merlin reports an expanded module type that's 5400 lines long). It has several sub-modules that reference each other heavily.
```
module Make(Config : Config.S) : S = struct
  module Config = Config
  module My_foo = Foo(Config)
  module My_bar = Bar(My_foo)
  
  (* ... many, many more things ... *)
end
```
`Config.S` is basically a record type with 18 small fields, but `Foo` and `Bar` are pretty substantial functors in their own right. Then the driver comes along and says:

```
module Implementations = struct
  module I1 = Make(Config1)
  module I2 = Make(Config2)
  module I3 = Make(Config3)
end

let run n =
  let module Implementation =
    (val
      match n with
      | 1 -> (module Implementations.I1 : S)
      | 2 -> (module Implementations.I2)
      | _ -> (module Implementations.I3))
  in
  ...
```

Up to `run`, all of this compiles fine: with aggressive inlining, the three functor calls in `Implementations` generate _a lot_ of code, but nothing we can't handle (in about 30 s). The `let module`, however, is trouble: It produces a join point whose argument type is `I1 ∨ I2 ∨ I3`, which is a three-way join over the _entire_ signature `S`. Worse, join operations produce expanded types that have no sharing: every reference to `Config` in `Foo` becomes a new record type with 18 fields, and then every reference to `My_foo` in `Bar` becomes a _giant_ new record type _in which `Config` has been expanded many times_. The result of all this (with `Foo` doing something similar on its own) is an OOM on a 96G machine.

(For reasons I haven't quite pinned down, the OOM only happens if the max inlining depth is at least 9, even though I'm running this with functor return types, so in principle the types should get huge even without inlining. My working theory is that inlining often causes functions to have more free variables, and thus bigger closures: a function where `I1` occurs free, for instance, might become one where `I1` and `I1.My_bar` _separately_ occur free, probably ~doubling the size of the fully-expanded type of the closure.)

It's not entirely clear what to do here. As a workaround, one can wrap each branch in `Sys.opaque_identity` (as it happens, the join isn't that useful anyway), but that's a messy hack for perfectly reasonable code. It should suffice for `Meet_and_join.join` to use memoization to preserve sharing, but its recursive calls take an environment that keeps changing, so it would have to be done carefully if it's possible at all. Of course, `join` could give up when its arguments get too large; that might be a bit brutal in cases like this, but I doubt it happens often that

  1. there are modules with enough sharing to explode like this, and
  2. they're large enough to eat this much RAM, _and_
  3. there would be much benefit from computing the join in full.

**Comment by lthls** [comment]
So if I understand correctly, the issue can be summarized  by the following example:
```
Join inputs:
left:
  res -> Block (tag = _, size = 2) : [ (=x); (=x) ]
  x -> head_x
right:
  res -> Block (tag = _, size = 2) : [ (=y); (=y) ]
  y -> head_y

Result:
  res -> Block (tag = _, size = 2) : [ head_x U head_y; head_x U head_y ]
```
It's easy to find a better result for this particular case (we could create a new variable `x_or_y` when we join the alias types for `x` and `y`, and reuse it each time we do the same join), but it's not trivial to decide when it makes sense to use this heuristic instead of the current one.
My intuition is that we could systematically create a new variable for each pair of equivalence classes we're joining, and at the end of the join replace all the ones that were only used once by their expanded heads. But I'd need some time to implement it properly.

**Comment by lukemaurer** [comment]
Yeah, that sounds reasonable. Here's a more-or-less minimal working example:
[eat_all_the_ram.ml](https://github.com/ocaml-flambda/flambda-backend/files/7907303/eat_all_the_ram.ml.txt)
